### PR TITLE
Add support for new 9.2 events

### PR DESCRIPTION
**Pull request by mpickering**
* EVENT_IPE - Information about the provenenace of info tables, used
with -hi profiling
* EVENT_MEM_RETURN - Statistics about how much memory the RTS thinks it
needs and how much it manages to return
* EVENT_BLOCKS_SIZE - The total size of the currently allocated blocks

### CODE REVIEW

**Comment by maoe** [comment]
Thank you! I guess I can take a closer look tonight.

Could you add a small example eventlog file that contains those new events with expected output to the [test suite](https://github.com/haskell/ghc-events/blob/5dfa473180f368214e488e0e9c63812c9c973fb2/test/Utils.hs#L3-L24)?

File: src/GHC/RTS/Events/Binary.hs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE TypeApplications #-}
```
Comment (maoe) [inline src/GHC/RTS/Events/Binary.hs:4]: This is causing CI for 7.10.3 to fail. Maybe we can just drop support for GHC 7.10.3.

**Comment by mpickering** [comment]
Added a test now and fixed the 7.10.3 build.

File: src/GHC/RTS/Events/Binary.hs [hunk 2]
```diff
@@ -813,6 +830,28 @@ heapProfParsers =
         ])
       (return ())
     return $! HeapProfCostCentre {..}
+  , VariableSizeParser EVENT_IPE $ do
+    payloadLen <- get :: Get Word16
+    itInfo <- get
+    itTableName <- getTextNul
+    itClosureDescText <- getTextNul
+    let itClosureDesc :: Int = read (T.unpack itClosureDescText)
```
Comment (maoe) [inline src/GHC/RTS/Events/Binary.hs:838]: nit: To avoid `String` conversion

```suggestion
    itClosureDesc <- either fail (return . fst) $ TR.decimal itClosureDescText
```
assuming `Data.Text.Read` is imported as `TR`.

File: src/GHC/RTS/EventTypes.hs [hunk 3]
```diff
@@ -220,6 +220,11 @@ data EventInfo
                        , parTotCopied :: {-# UNPACK #-}!Word64
                        , parBalancedCopied :: !(Maybe Word64)
                        }
+  | MemReturn          { heapCapset :: !Capset
+                       , current :: Int
+                       , needed :: Int
+                       , returned :: Int
```
Comment (maoe) [inline src/GHC/RTS/EventTypes.hs:226]: Why are these fields lazy?

File: src/GHC/RTS/Events/Binary.hs [hunk 4]
```diff
@@ -103,7 +104,7 @@ getEvent (EventParsers parsers) = do
      then return Nothing
      else do !evTime   <- get
              evSpec <- parsers ! fromIntegral etRef
-             return $ Just Event { evCap = undefined, .. }
+             return $ Just Event { evCap = Nothing, .. }
```
Comment (maoe) [inline src/GHC/RTS/Events/Binary.hs:107]: I don't know why it was undefined previously but why is this change needed?

File: src/GHC/RTS/Events/Binary.hs [hunk 5]
```diff
@@ -177,6 +178,17 @@ standardParsers = [
...
  )),
 
+ (FixedSizeParser EVENT_MEM_RETURN (sz_capset + 3*4) (do
+      heapCapset   <- get
+      currentw     <- get :: Get Word32
+      neededw      <- get :: Get Word32
+      returnedw    <- get :: Get Word32
+      return MemReturn{ current = fromIntegral currentw
+                       , needed = fromIntegral neededw
+                       , returned = fromIntegral returnedw
```
Comment (maoe) [inline src/GHC/RTS/Events/Binary.hs:188]: Why do we want to cast these `Word32`s to `Int`s here?

File: src/GHC/RTS/Events/Binary.hs [hunk 6]
```diff
@@ -188,6 +200,11 @@ standardParsers = [
       sizeBytes  <- get
       return HeapSize{..}
  )),
+ (FixedSizeParser EVENT_BLOCKS_SIZE (sz_capset + 8) (do  -- (heap_capset, blocks_size)
+      heapCapset <- get
+      blocksSize  <- get
+      return BlocksSize{..}
```
Comment (maoe) [inline src/GHC/RTS/Events/Binary.hs:206]: nit: 
```suggestion
      return $! BlocksSize{..}
```
to avoid unnecessary thunks.

File: include/EventLogFormat.h [hunk 7]
```diff
@@ -199,6 +201,7 @@
 #define EVENT_HEAP_BIO_PROF_SAMPLE_BEGIN   166
 #define EVENT_PROF_SAMPLE_COST_CENTRE      167
 #define EVENT_PROF_BEGIN                   168
+#define EVENT_IPE                          169
```
Comment (maoe) [inline include/EventLogFormat.h:204]: What does IPE stand for?

File: src/GHC/RTS/EventTypes.hs [hunk 8]
```diff
@@ -396,6 +404,13 @@ data EventInfo
                        , heapProfSrcLoc :: !Text
                        , heapProfFlags :: !HeapProfFlags
                        }
+  | InfoTableProv      { itInfo :: !Word64
```
Comment (maoe) [inline src/GHC/RTS/EventTypes.hs:407]: What does Prov stand for?

**Review by maoe — COMMENTED** [review]
Thanks for your contribution. Mostly looks good. I have some questions and nitpickings.

File: src/GHC/RTS/EventTypes.hs (hunk 8 above)
Comment (maoe) [reply src/GHC/RTS/EventTypes.hs:407]: It's Provenance I guess.

File: src/GHC/RTS/Events/Binary.hs (hunk 4 above)
Comment (mpickering) [reply src/GHC/RTS/Events/Binary.hs:107]: I was trying to print out this value at a point where it was undefined. I will revert it.

File: src/GHC/RTS/Events/Binary.hs (hunk 5 above)
Comment (mpickering) [reply src/GHC/RTS/Events/Binary.hs:188]: Yes good point.

File: include/EventLogFormat.h (hunk 7 above)
Comment (mpickering) [reply include/EventLogFormat.h:204]: `InfoTableProvenanceEntry`

File: src/GHC/RTS/Events/Binary.hs (hunk 6 above)
Comment (mpickering) [reply src/GHC/RTS/Events/Binary.hs:206]: Done.

**Comment by mpickering** [comment]
Updated now thanks to your comments.

File: include/EventLogFormat.h (hunk 7 above)
Comment (maoe) [inline include/EventLogFormat.h:204]: I see. Thanks.

File: src/GHC/RTS/Events/Binary.hs [hunk 9]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE CPP #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE ViewPatterns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
```
Comment (maoe) [inline src/GHC/RTS/Events/Binary.hs:4]: This extension is no longer needed.

File: src/GHC/RTS/Events/Binary.hs [hunk 10]
```diff
@@ -177,6 +179,17 @@ standardParsers = [
                        , ..}
  )),
 
+ (FixedSizeParser EVENT_MEM_RETURN (sz_capset + 3*4) (do
+      heapCapset   <- get
+      current      <- get :: Get Word32
+      needed       <- get :: Get Word32
+      returned     <- get :: Get Word32
+      return MemReturn{ current = current
```
Comment (maoe) [inline src/GHC/RTS/Events/Binary.hs:187]:
```suggestion
      return $! MemReturn{ current = current
```

**Review by maoe — CHANGES_REQUESTED** [review]

File: include/EventLogFormat.h (hunk 7 above)
Comment (maoe) [reply include/EventLogFormat.h:204]: I see. Thanks.

**Comment by maoe** [comment]
Otherwise LGTM

**Comment by mpickering** [comment]
Updated with your requested changes.

**Comment by maoe** [comment]
Looks perfect! Thank you.

**Review by maoe — APPROVED** [review]
